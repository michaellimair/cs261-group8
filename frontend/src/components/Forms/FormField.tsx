import {
  FormControl,
  FormLabel,
  Input,
  InputGroup,
  InputRightElement,
  Button,
  FormErrorMessage,
} from '@chakra-ui/react';
import React, {
  useState, useEffect, PropsWithChildren,
} from 'react';
import { ViewIcon, ViewOffIcon } from '@chakra-ui/icons';
import { useTranslation } from 'react-i18next';
import { IFieldProps } from 'customTypes/form';

const FormField = <T extends any>({
  name,
  autoComplete,
  register,
  error,
  required,
  label: formLabel,
  type = 'text',
  placeholder,
}: PropsWithChildren<IFieldProps<T>>) => {
  const [showPassword, setShowPassword] = useState<boolean>(false);
  const [fieldType, setFieldType] = useState<React.HTMLInputTypeAttribute>(type);
  const { t } = useTranslation();

  useEffect(() => {
    if (type !== 'password') {
      return;
    }
    if (showPassword) {
      setFieldType('text');
    } else {
      setFieldType('password');
    }
  }, [showPassword, type]);

  return (
    <FormControl id={name} data-testid={`${name}-formcontrol`} isRequired={required} mb={2} isInvalid={Boolean(error)}>
      <FormLabel data-testid={`${name}-label`}>{formLabel ?? t(name)}</FormLabel>
      <InputGroup>
        <Input
          type={fieldType}
          data-testid={name}
          autoComplete={autoComplete}
          {...register(name, {
            required: required ? t('no_blank', { field: formLabel ?? t(name) }) as any : undefined,
            valueAsNumber: type === 'number',
          })}
          placeholder={placeholder}
        />
        {type === 'password' && (
        <InputRightElement h="full" data-testid="show-password-toggle">
          <Button
            variant="ghost"
            data-testid="show-password-button"
            onClick={() => {
              setShowPassword((show) => !show);
            }}
          >
            {showPassword ? <ViewIcon data-testid="password-shown" /> : <ViewOffIcon data-testid="password-hidden" />}
          </Button>
        </InputRightElement>
        )}
      </InputGroup>
      <FormErrorMessage>
        {error}
      </FormErrorMessage>
    </FormControl>
  );
};

export default FormField;
