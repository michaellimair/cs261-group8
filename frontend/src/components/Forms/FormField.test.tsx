import { act, render, RenderResult } from '@testing-library/react';
import FormField from './FormField';

interface ISampleForm {
  field: string;
}

describe('components/FormField', () => {
  const fieldName: keyof ISampleForm = 'field';
  const autoComplete = 'email';
  const type: React.HTMLInputTypeAttribute = 'email';
  let result: RenderResult<typeof import('@testing-library/dom/types/queries'), HTMLElement>;
  let registerFn: jest.Mock;

  beforeEach(() => {
    registerFn = jest.fn();
    result = render(
      <FormField<ISampleForm>
        name={fieldName}
        type={type}
        autoComplete={autoComplete}
        register={registerFn}
      />,
    );
  });

  afterEach(() => {
    registerFn.mockReset();
    result.unmount();
  });

  it('renders correctly with the right set of attributes', async () => {
    const inputFormControl = result.queryByTestId(`${fieldName}-formcontrol`)!;
    expect(inputFormControl).not.toBeNull();
    expect(inputFormControl).toMatchSnapshot();

    const inputLabelField = inputFormControl.querySelector('label')!;
    expect(inputLabelField).toBeTruthy();
    expect(inputLabelField.innerHTML).toEqual(fieldName);

    const inputField = inputFormControl.querySelector('input')!;
    expect(inputField).toBeTruthy();
    expect(inputField.id).toEqual(fieldName);
    expect(inputField.getAttribute('autoComplete')).toEqual(autoComplete);
    expect(inputField.type).toEqual(type);
  });

  it('invokes the register function to register the form field for non-required fields', async () => {
    expect(registerFn).toHaveBeenCalledTimes(1);
    expect(registerFn).toHaveBeenCalledWith(fieldName, {
      required: undefined,
    });
  });

  it('invokes the register function to register the form field for required fields', async () => {
    result.unmount();
    registerFn.mockReset();
    result = render(
      <FormField<ISampleForm>
        name={fieldName}
        type={type}
        autoComplete={autoComplete}
        register={registerFn}
        required
      />,
    );

    expect(registerFn).toHaveBeenCalledTimes(1);
    expect(registerFn).toHaveBeenCalledWith(fieldName, {
      required: 'no_blank',
    });
  });

  it('displays the field error when there is an error', async () => {
    const errorText = 'This is a sample error text.';

    result.unmount();
    registerFn.mockReset();
    result = render(
      <FormField<ISampleForm>
        name={fieldName}
        type="password"
        autoComplete="given-name"
        register={registerFn}
        error={errorText}
      />,
    );

    const errorInField = await result.findByText(errorText);
    expect(errorInField).toBeTruthy();
  });

  it('displays a functional password toggle when the field type is password', () => {
    result.unmount();
    result = render(
      <FormField<ISampleForm>
        name={fieldName}
        type="password"
        autoComplete="given-name"
        register={jest.fn()}
      />,
    );

    const showPasswordToggle = result.queryByTestId('show-password-toggle')!;
    expect(showPasswordToggle).not.toBeNull();

    // Password will be hidden by default
    const showPassButton = result.queryByTestId('show-password-button')!;
    expect(showPassButton).not.toBeNull();

    const inputField = result.queryByTestId(fieldName)!;
    expect(inputField.getAttribute('type')).toBe('password');

    // Show password
    act(() => {
      showPassButton.click();
    });
    let passwordOffIcon: HTMLElement | null;
    let passwordOnIcon: HTMLElement | null;

    const passwordHiddenId = 'password-hidden';
    const passwordShownId = 'password-shown';
    passwordOffIcon = result.queryByTestId(passwordHiddenId);
    expect(passwordOffIcon).toBeNull();
    passwordOnIcon = result.queryByTestId(passwordShownId);
    expect(passwordOnIcon).not.toBeEmptyDOMElement();
    expect(inputField.getAttribute('type')).toBe('text');

    // Hide password again
    act(() => {
      showPassButton.click();
    });

    passwordOffIcon = result.queryByTestId(passwordHiddenId);
    expect(passwordOffIcon).not.toBeEmptyDOMElement();
    passwordOnIcon = result.queryByTestId(passwordShownId);
    expect(passwordOnIcon).toBeNull();
    expect(inputField.getAttribute('type')).toBe('password');
  });
});
