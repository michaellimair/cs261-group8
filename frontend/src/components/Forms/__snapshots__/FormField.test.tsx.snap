// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/FormField renders correctly with the right set of attributes 1`] = `
<div
  class="chakra-form-control css-wlea3r"
  data-testid="field-formcontrol"
  role="group"
>
  <label
    class="chakra-form__label css-uvho5g"
    data-testid="field-label"
    for="field"
    id="field-label"
  >
    customlabel
  </label>
  <div
    class="chakra-input__group css-4302v8"
  >
    <input
      autocomplete="email"
      class="chakra-input css-0"
      data-testid="field"
      id="field"
      placeholder="testplaceholder"
      type="email"
    />
  </div>
</div>
`;
